platform/android/android_input_handler.cpp

```diff
@@ -229,88 +224,118 @@ void AndroidInputHandler::process_touch(int p_event, int p_pointer, const Vector
+	if (p_source_mouse_relative) {
+		ev->set_position(hover_prev_pos);
+		ev->set_global_position(hover_prev_pos);
+	} else {
+		ev->set_position(mouse_event_info.pos);
+		ev->set_global_position(mouse_event_info.pos);
+		hover_prev_pos = mouse_event_info.pos;
+	}
+	ev->set_pressed(p_pressed);
+	MouseButton changed_button_mask = MouseButton(buttons_state ^ event_buttons_mask);
+
+	buttons_state = event_buttons_mask;
+
+	ev->set_button_index(_button_index_from_mask(changed_button_mask));
+	ev->set_button_mask(event_buttons_mask);
+	ev->set_double_click(p_double_click);
+	Input::get_singleton()->parse_input_event(ev);
+}
+
+void AndroidInputHandler::_release_mouse_event_info(bool p_source_mouse_relative) {
+	_parse_mouse_event_info(MouseButton::NONE, false, false, p_source_mouse_relative);
+	mouse_event_info.valid = false;
+}
+
+void AndroidInputHandler::process_mouse_event(int p_event_action, int p_event_android_buttons_mask, Point2 p_event_pos, Vector2 p_delta, bool p_double_click, bool p_source_mouse_relative) {
+	MouseButton event_buttons_mask = _android_button_mask_to_godot_button_mask(p_event_android_buttons_mask);
+	switch (p_event_action) {
 		case AMOTION_EVENT_ACTION_HOVER_MOVE: // hover move
 		case AMOTION_EVENT_ACTION_HOVER_ENTER: // hover enter
 		case AMOTION_EVENT_ACTION_HOVER_EXIT: { // hover exit
+			// https://developer.android.com/reference/android/view/MotionEvent.html#ACTION_HOVER_ENTER
 			Ref<InputEventMouseMotion> ev;
 			ev.instantiate();
 			_set_key_modifier_state(ev);
-			ev->set_position(p_pos);
-			ev->set_global_position(p_pos);
-			ev->set_relative(p_pos - hover_prev_pos);
+			ev->set_position(p_event_pos);
+			ev->set_global_position(p_event_pos);
+			ev->set_relative(p_event_pos - hover_prev_pos);
 			Input::get_singleton()->parse_input_event(ev);
-			hover_prev_pos = p_pos;
+			hover_prev_pos = p_event_pos;
 		} break;
-	}
-}
 
-void AndroidInputHandler::process_mouse_event(int input_device, int event_action, int event_android_buttons_mask, Point2 event_pos, float event_vertical_factor, float event_horizontal_factor) {
-	MouseButton event_buttons_mask = _android_button_mask_to_godot_button_mask(event_android_buttons_mask);
-	switch (event_action) {
-		case AMOTION_EVENT_ACTION_BUTTON_PRESS:
-		case AMOTION_EVENT_ACTION_BUTTON_RELEASE: {
-			Ref<InputEventMouseButton> ev;
-			ev.instantiate();
-			_set_key_modifier_state(ev);
-			if ((input_device & AINPUT_SOURCE_MOUSE) == AINPUT_SOURCE_MOUSE) {
-				ev->set_position(event_pos);
-				ev->set_global_position(event_pos);
-			} else {
-				ev->set_position(hover_prev_pos);
-				ev->set_global_position(hover_prev_pos);
-			}
-			ev->set_pressed(event_action == AMOTION_EVENT_ACTION_BUTTON_PRESS);
-			MouseButton changed_button_mask = MouseButton(buttons_state ^ event_buttons_mask);
+		case AMOTION_EVENT_ACTION_DOWN:
+		case AMOTION_EVENT_ACTION_BUTTON_PRESS: {
+			// Release any remaining touches or mouse event
+			_release_mouse_event_info();
```

Comment:
Why do you release on down? If I down primary and than down secondary the primary shouldn't be released, or am I missing somthing?

Reply:
This follows the approach used for the touch events and is necessary as the engine seems to get in a weird state when an input stream is started but not completed and another one is started in the middle of it.

During testing, I was getting into scenarios where input would stop responding, so to prevent those scenarios, I've extended the clearing logic used for touches to mouse events.

Reply:
It also prevents bugs where the `release` event is swallowed up upstream and not properly dispatched to the native layer by ensuring the ongoing input stream gets terminated when a new one starts.

Reply:
Does it matter where the inout comes from? But that sound a bug in godot as far as it looks, when I tested it last time I did it I didn't got any bug while doing things like that 😅

Reply:
It will nake it harder to just port to Android/chromebook from pc, as you consider holding 2 mouse buttons a pretty essential to a lot of games (shooters for example)

Reply:
I'm open to revisit this logic in a follow-up pr. I ran into the issue when I was testing and this approach addressed it.
Once we stabilize this logic (since it's new and may have bugs :) ), then we should revisit enabling support for that capability!